README
 _____                    __
|  _  \       __       __/  \
| | \  |     /  \__   /      \___
| |  | | ___/ _ __ \_/_  _ __  __\_
| |  |  / _ \| '_ \/ _ \| '__|/  _ |
| |_/  |  __/| | |  |_| | |  ' (_| |
|_____/ \___||_| |_\___/|_|   \__,_|_

          Statistics for IRC Networks
          http://www.denorastats.org/

-----------------
TABLE OF CONTENTS
-----------------
    1. Credits
    2. Presentation
    3. Installation
    4. Command line options
    5. Contact and support

1. CREDITS
----------

    Denora is based on the original code of Anope 1.7.6 by Anope Team,
    and based on the original code of Thales 1.0 by Lucas Nussbaum.

	Please see docs/CREDITS for a full list of credits.

    This program is free but copyrighted software; see the file docs/COPYING
    for details.


2. PRESENTATION
---------------

    Denora is a set of Stats Services for IRC networks that allows IRC
    network admins to get a complete and comprehensive overview of their
    IRC network. Denora exports this information to MySQL, HTML XML and
    flatfile databases.
    
    Denora's StatServ pseudo client, provides access to commands that
    allow to manage and view the network stats.

    Denora Stats coupled with phpDenora[1] or MagIRC[1] is the most powerful tool
    for generating statistics for IRC.

    Denora currently works with the following IRCds:

        * Bahamut 1.4 and 1.8
        * Charybdis 1.0 - 3.x
        * Inspircd 1.1.x - 2.0.x
        * IRCU 2.10.11 - 2.10.12
        * Nefarious IRCu 1.3.x - 2.0.x
        * Ratbox 2.0.6 - 3.0.x
        * ScaryNet IRCu 2.10.11 - 2.10.12
        * Unreal 3.2 beta19 or later

    Denora could also work with some of the deamons derivated by the ones 
    listed above.

    [1] http://www.denorastats.org/


3. INSTALLATION
---------------

	For a Quick Start guide on Unix or Windows systems see the QUICKSTART.txt file.
    See the docs/INSTALL file for detailed instructions for Unix-based systems.


4. COMMAND LINE OPTIONS
-----------------------
    Alternatively, any of the following command-line options can be specified
    in order to replace the default configuration:

        -remote server[:port]    Connect to the specified server
        -local host  -or-        Connect from the specified address (e.g.
               [host]:[port]     for multihomed servers)
        -name servername         Our server name (e.g. stats.some.net)
        -desc string             Description of us (e.g. SomeNet Statistics)
        -user username           Username for Stats' nicks (e.g. stats)
        -host hostname           Hostname for Stats' nicks (e.g. nomadinc.net)
        -dir directory           Directory containing Denora' data files
                                 (e.g. /usr/local/lib/stats)
        -log filename            Denora's log filename (e.g. stats.log)
        -update secs             How often to update databases (in seconds)

    Additionally, the following command-line options can be used to modify
    the behavior of Denora:

        -debug          Enable debugging mode
        -sql            Enable debugging of sql queries
        -version        Display the version of Denora
        -socketdebug    Display socket write/read info only useful if you want
                        to debug connection problems.
        -protocoldebug  Display protocol debug information, very helpful in 
                        supporting ircds.

    Denora also has a feature to crypt passwords to be used in the config file,
    to encrypt your password use the following command-line option:
        -mkpass=passwd  Encrypt a password


5. CONTACT AND SUPPORT
----------------------

    * If you read the documentation carefully, which you also can find on our wiki
      (http://wiki.denorastats.org/) and didn't find the answer to your
      question, feel free to post on the website forums or join our IRC channel
      (irc.denorastats.org #denora). Be as precise as possible when asking a question,
      because we have no extraordinary powers and can't guess things if they aren't
      provided. The more precise you are, the sooner you'll be likely to get an answer.

    * If you think you found a bug, add it to the bug tracking system on our
      website (http://bugs.anope.org/) and - again - be as
      precise as possible. Also say whether the bug happens always or under what
      circumstances, and anything that could be useful to track your bug down.
      If you wrote a patch, send it over.

    * Support is also available via our forums located at
      http://forum.anope.org/


Changes
Denora v1.5.1-git
-------------
Additions
13/12/09 : Support for Unreal User Mode +I / Channel mode +Z
14/01/09 : Added new link list system, which is far faster then the old one.

Changes
14/01/08 : Rewrote the UID tracking to be in link list.
13/12/11 : Restructored documentation, added html based documentation

Fixes
14/03/06 : Fixed crash caused by uid lookup after nick change
14/01/23 : Fixed some incorrect buffer sizes
14/01/10 : Fixed crashes with servers that use UIDs.
14/01/02 : Statserv export servers matches StatServ export all server block 
13/12/09 : Fixed issue where do_sjoin() did not create the channel before trying to set modes
13/12/02 : Some fixes to compile on Minix (not a 100% yet)
13/11/29 : Cleans out UID struct on user disconnecting
13/11/29 : Checks if server before calling nickisservices()
13/11/27 : Fixed crash when crypt.h is missing but the lib is not

Removals
14/01/23 : Removed legacy defines from the ADNS days

Developers
Denora Team <dev@denorastats.org>

Denora v1.5.0
-------------

Additions
13/11/22 Exclude now supports wildcard matching.
13/11/15 From command line -v (alias of -version) and -? (alias of -help) been added
13/11/04 Full support for ircu and nefarious channel mode +A/U
13/11/04 ModuleCreateConfigDirective() cleaner way to pass to moduleGetConfigDirective()
13/10/23 Adds m4 directory, the configure files are getting nasty, time to move macros where
         they belong.
13/10/21 Added regex.h to the build, core and modules can now use this to pattern 
         match with regular expressions. 
13/01/13 Added support for ircu's channel mode +R (was missing)

Changes
13/11/22 Removed mypasql.c and mypasql.def neither have been used in a long time.
13/11/22 Removed dmalloc, when actually hooked up correctly denora crashed using it.
13/11/20 base64_decode() removed last parameter of length, was never used.
13/11/18 Not finding mysql_config no longer causes configure to exit on an error, just throw a warning
13/11/15 Denora no longer says "Service not started" if the config file is missing, it gives a
         better error.
13/11/11 Moved do_umode() hack from 1.4.0 to the protocol like it should have been.
13/11/06 denora.sql no longer includes ison table data, all this is move to the ircd based
         sql files to remove extra columns that we will never set on that ircd.
13/11/03 ModData functions require you to call it with your modules name, since moddata structs
         were broken up until this point it shouldn't be a huge deal.
13/10/24 Removes some future features so 1.5 can roll
13/10/24 Removes Postgre from the code, was never used or really compiled for, maybe in the futre
13/10/22 Removed configure searches to embed perl in the code
13/10/17 Removed USE_RDB define from config, this was legacy Anope code for when rdb was an option.
13/05/31 Improved handling of KILL messages on InspIRCd
13/05/16 Do not free unallocated memory in InspIRCd protocol

12/10/04 Cleanup and hopefully less leaks etc
12/08/23 db_getserver() now escapes servername itself where needed
12/08/05 Threading is now disabled by default (it's only used for mysql and the current implementation
	 has serious issues)
12/07/26 Set country to server's country if connection is localhost
12/07/26 Use geoip's nice ntop/pton wrapper and depend on the existance of the functions, not the OS
12/07/26 More cleanup and rewrites in src/users.c src/admin.c src/bans.c src/base64.c src/GeoIP.c
         src/ChanBanMode.c src/ChanMode.c src/channels.c
12/07/25 More cleanup and rewrite
12/07/24 Lots of cleanup in src/db.c
12/07/23 Rewrote db file handling a little
12/07/23 Automode handled in denora_cmd_join now, removed from other obsolete places (fixed #1433)
12/07/23 Updated sql specific code, more preparations for threadin
12/07/18 Fixed persistent channel support by actually setting the modes in the protocols.
12/07/18 Applied denora_automode to several protocols
12/07/18 Moved automode handler into a separate void and marked some todo's regarding it for later
12/07/18 Added automode handler for logchannel into inspircd11
12/07/17 p10 wants 391 replies (RPL_TIME) in another way. 'Fixed', just need to calc offset somehow.
12/07/17 Don't call events if (umode = o && ((add && was_oper) || (!add && !was_oper)) (should fix #1264)
12/07/17 Updated README
12/07/15 used #ifdef USE_MYSQL in modules seen/lastspoke/ircd_ports (should fix #1429)
12/07/15 File permission detection in mysql_backup.c (should fix #1127)
12/07/15 Added automode for logchan in bahamut
12/07/15 Adding automode for logchan in inspircd12.c (should fix #1418)
12/07/15 Don't init something to NULL after free if you gonna sstrdup something over it anyways.
12/07/15 More free/sstrdup usage / Cleaned up seen.c and made it ^M free
12/07/14 Revoked change to DenoraFree() and removed that macro entirely
12/07/09 Updated ScaryNet Protocol
12/07/06 Changed delete_user to properly and warning free cleanup the channels
12/07/05 Changed a lot of free()'s into DenoraFree() in order to prevent double free()'s
12/06/29 Undone a few changes as I might see issues
12/06/27 Commented out a little more at src/db.c
12/06/27 Bumped to 1.5.0-git-501
12/06/26 Use inet_ntop for decode_ip function
12/06/26 Reduced mysql_backup.c by 27%
12/04/25 Channel names no longer forced to lowercase (#1403)
12/04/24 Updated GeoIP databases
12/04/11 Added 'cron' function to denorarc for crontab usage 
12/04/11 Updated ScaryNet protocol
11/05/04 r435 Fixed formatting of docs/EVENTS
11/04/10 r428 documented the -protocoldebug option in the -help output
11/04/10 r428 made denora accept options that start with --
11/02/11 r419 Updated Denora icon
11/02/11 r416 Updated Windows installer

Fixes
13/11/22 Denora links with hybrid using TS6 enabled.
13/11/22 Compiling with clang on bsd systems.
13/11/20 Fixed compiler warnings created by fixing stuff for macosx, so now it will build nicely on
         slackware.

src/md5.c
/*
 * This code implements the MD5 message-digest algorithm.
 * The algorithm is due to Ron Rivest.  This code was
 * written by Colin Plumb in 1993, no copyright is claimed.
 * This code is in the public domain; do with it what you wish.
 *
 * Equivalent code is available from RSA Data Security, Inc.
 * This code has been tested against that, and is equivalent,
 * except that you don't need to include two pages of legalese
 * with every copy.
 *
 * To compute the message digest of a chunk of bytes, declare an
 * MD5Context structure, pass it to MD5Init, call MD5Update as
 * needed on buffers full of bytes, and then call MD5Final, which
 * will fill a supplied 16-byte array with the digest.
 */

#include "denora.h"

#ifndef HIGHFIRST
/** Bit-reverse bytes in a buffer. */
#define byteReverse(buf, len)	/* Nothing */
#else
static void byteReverse(unsigned char *buf, unsigned longs);

#ifndef ASM_MD5
/*
 * Note: this code is harmless on little-endian machines.
 */
static void byteReverse(unsigned char *buf, unsigned longs)
{
	uint32 t;
	do {
		t = (uint32) ((unsigned) buf[3] << 8 | buf[2]) << 16 |
		    ((unsigned) buf[1] << 8 | buf[0]);
		*(uint32 *) buf = t;
		buf += 4;
	} while (--longs);
}

#endif
#endif

#ifndef ASM_MD5
/* The four core functions - F1 is optimized somewhat */
/* #define F1(x, y, z) (x & y | ~x & z) */
/** Helper function for first round of MD5. */
#define F1(x, y, z) (z ^ (x & (y ^ z)))
/** Helper function for second round of MD5. */
#define F2(x, y, z) F1(z, x, y)
/** Helper function for third round of MD5. */
#define F3(x, y, z) (x ^ y ^ z)
/** Helper function for fourth round of MD5. */
#define F4(x, y, z) (y ^ (x | ~z))

/** Step function for each round of MD5 */
#define MD5STEP(f, w, x, y, z, data, s) \
	( w += f(x, y, z) + data,  w = w<<s | w>>(32-s),  w += x )
#endif

/** Iniitalize MD5 context.
 * @param[out] ctx MD5 context to initialize.
 */
void MD5Init(struct MD5Context *ctx);
void MD5Init(struct MD5Context *ctx)
{
	ctx->buf[0] = 0x67452301U;
	ctx->buf[1] = 0xefcdab89U;
	ctx->buf[2] = 0x98badcfeU;
	ctx->buf[3] = 0x10325476U;

	ctx->bits[0] = 0;
	ctx->bits[1] = 0;
}

/** Perform the core MD5 update steps to update a 128-bit hash value
 * with 512 bits of input data.
 * @param[in,out] buf Hash value.
 * @param[in] in Input buffer.
 */
void MD5Transform(uint32 buf[4], uint32 const in[16]);
void MD5Transform(uint32 buf[4], uint32 const in[16])
{
	register uint32 a, b, c, d;

	a = buf[0];
	b = buf[1];
	c = buf[2];
	d = buf[3];

	MD5STEP(F1, a, b, c, d, in[0] + 0xd76aa478U, 7);
	MD5STEP(F1, d, a, b, c, in[1] + 0xe8c7b756U, 12);
	MD5STEP(F1, c, d, a, b, in[2] + 0x242070dbU, 17);
	MD5STEP(F1, b, c, d, a, in[3] + 0xc1bdceeeU, 22);
	MD5STEP(F1, a, b, c, d, in[4] + 0xf57c0fafU, 7);
	MD5STEP(F1, d, a, b, c, in[5] + 0x4787c62aU, 12);
	MD5STEP(F1, c, d, a, b, in[6] + 0xa8304613U, 17);
	MD5STEP(F1, b, c, d, a, in[7] + 0xfd469501U, 22);
	MD5STEP(F1, a, b, c, d, in[8] + 0x698098d8U, 7);
	MD5STEP(F1, d, a, b, c, in[9] + 0x8b44f7afU, 12);
	MD5STEP(F1, c, d, a, b, in[10] + 0xffff5bb1U, 17);
	MD5STEP(F1, b, c, d, a, in[11] + 0x895cd7beU, 22);
	MD5STEP(F1, a, b, c, d, in[12] + 0x6b901122U, 7);
	MD5STEP(F1, d, a, b, c, in[13] + 0xfd987193U, 12);
	MD5STEP(F1, c, d, a, b, in[14] + 0xa679438eU, 17);
	MD5STEP(F1, b, c, d, a, in[15] + 0x49b40821U, 22);

	MD5STEP(F2, a, b, c, d, in[1] + 0xf61e2562U, 5);
	MD5STEP(F2, d, a, b, c, in[6] + 0xc040b340U, 9);
	MD5STEP(F2, c, d, a, b, in[11] + 0x265e5a51U, 14);
	MD5STEP(F2, b, c, d, a, in[0] + 0xe9b6c7aaU, 20);
	MD5STEP(F2, a, b, c, d, in[5] + 0xd62f105dU, 5);
	MD5STEP(F2, d, a, b, c, in[10] + 0x02441453U, 9);
	MD5STEP(F2, c, d, a, b, in[15] + 0xd8a1e681U, 14);
	MD5STEP(F2, b, c, d, a, in[4] + 0xe7d3fbc8U, 20);
	MD5STEP(F2, a, b, c, d, in[9] + 0x21e1cde6U, 5);
	MD5STEP(F2, d, a, b, c, in[14] + 0xc33707d6U, 9);
	MD5STEP(F2, c, d, a, b, in[3] + 0xf4d50d87U, 14);
	MD5STEP(F2, b, c, d, a, in[8] + 0x455a14edU, 20);
	MD5STEP(F2, a, b, c, d, in[13] + 0xa9e3e905U, 5);
	MD5STEP(F2, d, a, b, c, in[2] + 0xfcefa3f8U, 9);
	MD5STEP(F2, c, d, a, b, in[7] + 0x676f02d9U, 14);
	MD5STEP(F2, b, c, d, a, in[12] + 0x8d2a4c8aU, 20);

	MD5STEP(F3, a, b, c, d, in[5] + 0xfffa3942U, 4);
	MD5STEP(F3, d, a, b, c, in[8] + 0x8771f681U, 11);
	MD5STEP(F3, c, d, a, b, in[11] + 0x6d9d6122U, 16);
	MD5STEP(F3, b, c, d, a, in[14] + 0xfde5380cU, 23);
	MD5STEP(F3, a, b, c, d, in[1] + 0xa4beea44U, 4);
	MD5STEP(F3, d, a, b, c, in[4] + 0x4bdecfa9U, 11);
	MD5STEP(F3, c, d, a, b, in[7] + 0xf6bb4b60U, 16);
	MD5STEP(F3, b, c, d, a, in[10] + 0xbebfbc70U, 23);
	MD5STEP(F3, a, b, c, d, in[13] + 0x289b7ec6U, 4);
	MD5STEP(F3, d, a, b, c, in[0] + 0xeaa127faU, 11);
	MD5STEP(F3, c, d, a, b, in[3] + 0xd4ef3085U, 16);
	MD5STEP(F3, b, c, d, a, in[6] + 0x04881d05U, 23);
	MD5STEP(F3, a, b, c, d, in[9] + 0xd9d4d039U, 4);
	MD5STEP(F3, d, a, b, c, in[12] + 0xe6db99e5U, 11);
	MD5STEP(F3, c, d, a, b, in[15] + 0x1fa27cf8U, 16);
	MD5STEP(F3, b, c, d, a, in[2] + 0xc4ac5665U, 23);

	MD5STEP(F4, a, b, c, d, in[0] + 0xf4292244U, 6);
	MD5STEP(F4, d, a, b, c, in[7] + 0x432aff97U, 10);
	MD5STEP(F4, c, d, a, b, in[14] + 0xab9423a7U, 15);
	MD5STEP(F4, b, c, d, a, in[5] + 0xfc93a039U, 21);
	MD5STEP(F4, a, b, c, d, in[12] + 0x655b59c3U, 6);
	MD5STEP(F4, d, a, b, c, in[3] + 0x8f0ccc92U, 10);
	MD5STEP(F4, c, d, a, b, in[10] + 0xffeff47dU, 15);
	MD5STEP(F4, b, c, d, a, in[1] + 0x85845dd1U, 21);
	MD5STEP(F4, a, b, c, d, in[8] + 0x6fa87e4fU, 6);
	MD5STEP(F4, d, a, b, c, in[15] + 0xfe2ce6e0U, 10);
	MD5STEP(F4, c, d, a, b, in[6] + 0xa3014314U, 15);
	MD5STEP(F4, b, c, d, a, in[13] + 0x4e0811a1U, 21);
	MD5STEP(F4, a, b, c, d, in[4] + 0xf7537e82U, 6);
	MD5STEP(F4, d, a, b, c, in[11] + 0xbd3af235U, 10);
	MD5STEP(F4, c, d, a, b, in[2] + 0x2ad7d2bbU, 15);
	MD5STEP(F4, b, c, d, a, in[9] + 0xeb86d391U, 21);

	buf[0] += a;
	buf[1] += b;
	buf[2] += c;
	buf[3] += d;
}

/** Update MD5 context with data from a buffer.
 * @param[in,out] ctx MD5 context to operate on.
 * @param[in] buf Input buffer.
 * @param[in] len Number of bytes in input buffer.
 */
void MD5Update(struct MD5Context *ctx, unsigned const char *buf, unsigned len);
void MD5Update(struct MD5Context *ctx, unsigned const char *buf, unsigned len)
{
	uint32 t;

	/* Update bitcount */

	t = ctx->bits[0];
	if ((ctx->bits[0] = t + ((uint32) len << 3)) < t)
		ctx->bits[1]++;	/* Carry from low to high */
	ctx->bits[1] += len >> 29;

	t = (t >> 3) & 0x3f;	/* Bytes already in shsInfo->data */

	/* Handle any leading odd-sized chunks */

	if (t) {
		unsigned char *p = (unsigned char *) ctx->in + t;

		t = 64 - t;
		if (len < t) {
			memcpy(p, buf, len);
			return;
		}
		memcpy(p, buf, t);
		byteReverse(ctx->in, 16);
		MD5Transform(ctx->buf, (uint32 *) ctx->in);
		buf += t;
		len -= t;
	}
	/* Process data in 64-byte chunks */

	while (len >= 64) {
		memcpy(ctx->in, buf, 64);
		byteReverse(ctx->in, 16);
		MD5Transform(ctx->buf, (uint32 *) ctx->in);
		buf += 64;
		len -= 64;
	}

	/* Handle any remaining bytes of data. */

	memcpy(ctx->in, buf, len);
}

/** Perform final steps of MD5 hash.
 * @param[out] digest Receives output hash value.
 * @param[in,out] ctx MD5 context to finalize.
 */
void MD5Final(unsigned char digest[16], struct MD5Context *ctx);
void MD5Final(unsigned char digest[16], struct MD5Context *ctx)
{
	unsigned count;
	unsigned char *p;

	/* Compute number of bytes mod 64 */
	count = (ctx->bits[0] >> 3) & 0x3F;

	/* Set the first char of padding to 0x80.  This is safe since there is
	   always at least one byte free */
	p = ctx->in + count;
	*p++ = 0x80;

	/* Bytes of padding needed to make 64 bytes */
	count = 64 - 1 - count;

	/* Pad out to 56 mod 64 */
	if (count < 8) {
		/* Two lots of padding:  Pad the first block to 64 bytes */
		memset(p, 0, count);
		byteReverse(ctx->in, 16);
		MD5Transform(ctx->buf, (uint32 *) ctx->in);

		/* Now fill the next block with 56 bytes */
		memset(ctx->in, 0, 56);
	} else {
		/* Pad block to 56 bytes */
		memset(p, 0, count - 8);
	}
	byteReverse(ctx->in, 14);

	/* Append length in bits and transform */
	((uint32 *) ctx->in)[14] = ctx->bits[0];
	((uint32 *) ctx->in)[15] = ctx->bits[1];

	MD5Transform(ctx->buf, (uint32 *) ctx->in);
	byteReverse((unsigned char *) ctx->buf, 4);
	memcpy(digest, ctx->buf, 16);
	memset(ctx, 0, sizeof(*ctx));	/* In case it's sensitive */
}

char *md5(const char *str) {
        int n;
        int length = strlen(str);
        MD5_CTX c;
        unsigned char digest[16];
        char *out = (char*)malloc(33);

        MD5Init(&c);
        while (length > 0) {
                if (length > 512)
                        MD5Update(&c, (const unsigned char *)str, 512);
                else
                        MD5Update(&c, (const unsigned char *)str, length);
                length -= 512;
                str += 512;
        }
        MD5Final(digest, &c);

        for (n = 0; n < 16; ++n)
                ircsnprintf(&(out[n*2]), 16*2, "%02x", (unsigned int)digest[n]);

        memset(digest, 0, sizeof digest);
        return out;
}

